src/proxy/connection_manager.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::proxy::error;
+use crate::rbac::Connection;
+use crate::state::DemandProxyState;
+use drain;
+use std::collections::HashMap;
+use std::sync::{atomic::AtomicUsize, Arc};
+use tokio::sync::watch;
+use tokio::sync::RwLock;
+use tracing::{debug, info};
+
+struct ConnectionDrain {
+    tx: drain::Signal,
+    rx: drain::Watch,
+    count: AtomicUsize,
+}
+
+impl ConnectionDrain {
+    fn new() -> Self {
+        let (tx, rx) = drain::channel();
+        let count = AtomicUsize::new(1);
+        ConnectionDrain { tx, rx, count }
+    }
+
+    /// drain drops the internal reference to rx and then signals drain on the tx
+    // always inline, this is for convenience so that we don't forget to drop the rx but there's really no reason it needs to grow the stack
+    #[inline(always)]
+    async fn drain(self) {
+        drop(self.rx); // very important, drain cannont complete if there are outstand rx
+        self.tx.drain().await;
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionManager {
```

Comment:
We already have a "pool" tracking connections. It seems like there, ideally, would be a single data structure tracking connections. I guess Pool tracks outer HBONE connections while this tracks inner user connections, though. Also I don't see a feasible way to use 1 structure anyways, so mostly hypothetical.

Reply:
It's a good thought. I couldn't work out how to do it either though.